src/ts/plugins/server/GitlabHostingConnector.ts

```diff
@@ -0,0 +1,109 @@
+  connectorId = 'gitlab_hosting'
+  displayName = 'Gitlab hosting'
+  connectorType = ConnectorType.HOSTING
+
+  constructor( config: ServerConfig, opts: Partial<GitlabOptions>) {
+    super (config, opts)
+    /* public directory for standard gitlab pages */
+    this.options.assetsFolder= 'public'
+  }
+
+  async publish(session: GitlabSession, websiteId: WebsiteId, files: ConnectorFile[], {startJob, jobSuccess, jobError}: JobManager): Promise<JobData> {
+    const job = startJob(`Publishing to ${this.displayName}`) as PublicationJobData
+    job.logs = [[`Publishing to ${this.displayName}`]]
+    job.errors = [[]]
+    /* Configuration file .gitlab-ci.yml contains template for plain html Gitlab pages*/
+    const pathYml = '.gitlab-ci.yml'
+    const contentYml = 
+    `# This file is a template, and might need editing before it works on your project.
+    # To contribute improvements to CI/CD templates, please follow the Development guide at:
+    # https://docs.gitlab.com/ee/development/cicd/templates.html
+    # This specific template is located at:
+    # https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/ci/templates/Pages/HTML.gitlab-ci.yml
+    
+    # Full project: https://gitlab.com/pages/plain-html
+    image: busybox
+    
+    pages:
+      stage: deploy
+      environment: production
+      script:
+        - echo "The site will be deployed to $CI_PAGES_URL"
+      artifacts:
+        paths:
+          - public
+      rules:
+        - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH`
+    console.log(session)
+        try {
+      await this.readFile(session, websiteId, pathYml)
+    } catch (e) {
+      // If the file .gitlab-ci.yml does not exist, create it, otherwise do nothing (do not overwriting existing one)
+      if (e.statusCode === 404 || e.httpStatusCode === 404 || e.message.endsWith('A file with this name doesn\'t exist')) {
+        await this.createFile(session, websiteId, pathYml, contentYml)
+      }
+    }
+
+    /* Get Url Gitlab Pages */
+    const response = await this.callApi(session, `api/v4/projects/${websiteId}/pages`, 'GET')
+    const publicationUrl = response.url
+
+    /* publishing all files for website*/
+    await this.writeAssets(session, websiteId, files, async ({status, message}) => {
+      // Update the job status
+      job.status = status
+      job.message = message
+      job.logs[0].push(message)
+      if(status === JobStatus.SUCCESS) {
+        jobSuccess(job.jobId, '<a href="' + publicationUrl + '" target="_blank">' + publicationUrl + '<a>')
+      } else if(status === JobStatus.ERROR) {
+        job.errors[0].push(message)
+        jobError(job.jobId, message)
+      }
+    })
+    return job
+  }
+
+  async getUrl(session: GitlabSession, websiteId: WebsiteId): Promise<string> {
+    const id = websiteId
+    const filePath = join(this.options.domain, id, 'index.html')
+    const fileUrl = new URL(filePath, 'file://')
```

Comment:
This doesn't make sense for the gitlab connector
Please check in silex code base where this getUrl method is called? Then we see what to return

Reply:
I don't understand, it seems to be used in connector.ts and DownloadConnector.ts ?

Reply:
this is where it is used
https://github.com/silexlabs/Silex/blob/549bf7b6b2419113718b98b6ab97d0280252e871/src/ts/server/api/publicationApi.ts#L140

so after publication it will send an url to the client side, which gets it here
https://github.com/silexlabs/Silex/blob/549bf7b6b2419113718b98b6ab97d0280252e871/src/ts/client/grapesjs/PublicationManager.ts#L294

but then it is never used anywhere :face_with_diagonal_mouth: 

we could use this URL to display a big button "Go to your website" but this will be another feature for later

so for now, you should return the URL of the website. Maybe move the code which calls gitlab api there?

Reply:
Yes, clearer, I'll do that